src/Compiler/Service/IncrementalBuild.fs

```diff
@@ -1269,6 +1132,11 @@ type IncrementalBuilder(initialState: IncrementalBuilderInitialState, state: Inc
             do! setCurrentState currentState cache ct
         }
 
+    let checkFileTimeStampsSynchronously cache =
+        checkFileTimeStamps cache
+        |> Async.AwaitNodeCode
+        |> Async.RunSynchronously
```

Comment:
@majocha I started seeing some UI hangs recently in VS, specifically when finding references. Dump of the process shows that it's waiting on a task that involves this particular `RunSynchronously`. In this case checking project stamp through `IProjectReference`. 

Since this is on some paths that can be hit from different jobs quite frequently in parallel (when I was finding references semantic classification, document highlights and quick info services were also probably running), we might need to do something about it.

I saw that `IProjectReference.TryGetLogicalTimeStamp` is not async, but maybe we could just make it async? Are there any other reasons we need this to be synchronous?

Reply:
This is synchronous to keep type signature of call sites, only because it was all synchronous before. IIRC there was a temporary state computed instead of currentState, sometimes duplicating work.
I'll take a look. If this can't be made async we can try reverting this back to computing temp state.

Reply:
@0101 I think the problem is that currentState is guarded by semaphore while the tempState was not.

Reply:
Low cost attempt to fix this: #15146

Reply:
Thanks for looking into it!

Reply:
I think there may be something else going on here, too.
updating state in IncrementalBuilder should be instant when there are no actual changes and also super fast when there are changes. It just updates the graph, which is cold - won't do any work until requested to.

Reply:
There might, but avoiding `RunSynchronously` within async operations is still a good rule to keep if we can. 

I don't really get why it got stuck on this call. The dump shows UI thread waiting on this, which is being blocked by something - not sure what. And it originates from `IFSharpDocumentDiagnosticAnalyzer.AnalyzeSemanticsAsync`. 

![image](https://user-images.githubusercontent.com/217092/234811400-480e9e8a-7c8e-4e01-8d1d-54db2e85529f.png)

But I only ever saw the hangs when finding references, so there might be some weird interaction between those. 

I'm running now with your update so let's see if I see it again.

Reply:
I tried removing the semaphore completely, and it still blocks UI quite a lot in find references, at leas on my meager PC.

Reply:
I also experience this on 17.6.0 Preview 5.0, which I think does not include this PR? I kinda assumed search in VS was always causing UI lags a bit.

Reply:
Yeah some lags were always there, but with this my VS was completely frozen for about 10 minutes.

Reply:
Ah, I see, a total hang that's completely different.

Reply:
And I've now seen it again - with this already removed :( 

Though it was slightly different, the result was still the same, frozen VS. 

But this time it also ate all my memory. And the dump of the process shows 18 typechecks running and UI waiting for `AddSyntacticClassificationsAsync`.

![image](https://user-images.githubusercontent.com/217092/234837000-1b6d9bfb-82a6-42a8-8b24-ab4a195a8327.png)

So now I'm wondering if it could be also related to:
https://github.com/dotnet/fsharp/blob/9627d3373ee6c4c7dbb09295f5c8230cde303eea/vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs#L109

Although that's been there for a while and I haven't seen these issues until very recently.

Reply:
But maybe with parsing now inside the BuildGraph it's acting up.

Reply:
But also now it should be quite easy to make it async - I'm going to try that.

Reply:
I have a very faint suspicion, because I experienced weird behaviour recently, but in CI. I'm probably worng but maybe it's the same thing? Find references is very parallel from the get go with all the Task.WhenAll and so happens that this GetHashCode() here:
https://github.com/dotnet/fsharp/blob/9627d3373ee6c4c7dbb09295f5c8230cde303eea/vsintegration/src/FSharp.Editor/LanguageService/FSharpProjectOptionsManager.fs#L331
 is very susceptible to collisions when you create many at the same time and we use it for identity. This is grasping at straws but I'd put `None` there just to try if it helps.

Reply:
Hmm, I'm not even sure what's the point of that. If we have 2 project options where everything is the same except stamp - why do we care which one we use?

Reply:
The situation I had on mind is different project options getting the same stamp, corrupting service caches. But this is very far fetched.

Is there a way to try to trigger this bug? Just spam shift-f12 in VisualFSharp.sln?

Reply:
Oh I see, I read the description of `Stamp` wrong. But then I would expect just flat out errors or crashes if that happened. 

I think best chances to reproduce this are shift-F12 closely before or after changing some related files, or maybe newly opening them.